pkg/edgeview/src/basics.go

```diff
@@ -0,0 +1,608 @@
+		if _, err := os.Stat("/config"); err != nil {
+			return
+		}
+	}
+
+	retStr, err := runCmd("ip -4 -br a | grep UP", false, false)
+	if err == nil {
+		lines := splitLines(retStr)
+		ips := []string{}
+		for _, l := range lines {
+			if strings.HasPrefix(l, "bn") {
+				continue
+			}
+			//fmt.Printf("l: %s\n", l)
+			words := strings.Fields(l)
+			n := len(words)
+			ipaddrs := strings.Split(words[n-1], "/")
+			ips = append(ips, ipaddrs[0])
+		}
+		if myEvEndPoint != "" {
+			fmt.Printf("Device IPs: %v; Endpoint IP %s\n", ips, myEvEndPoint)
+		} else {
+			fmt.Printf("Device IPs: %v\n", ips)
+		}
+	}
+	retStr, err = runCmd("cat /persist/status/uuid", false, false)
+	if err == nil {
+		fmt.Printf("  UUID: %s", retStr)
+	}
+	retStr, err = runCmd("cat /config/server", false, false)
+	if err == nil {
+		var printed bool
+		conts := strings.Split(retStr, "zedcloud.")
+		if len(conts) == 2 {
+			cont2s := strings.Split(conts[1], ".zededa.net")
+			if len(cont2s) == 2 {
+				cluster := cont2s[0]
+				controller := strings.Replace(retStr, cluster, "\033[0;93m"+cluster+"\033[0m", 1)
+				fmt.Printf("  Controller: %s", controller)
+				printed = true
+			}
+		}
+		if !printed {
+			fmt.Printf("  Controller: %s", retStr)
+		}
+	}
+	retStr, err = runCmd("cat /run/eve-release && echo -n ', ' && cat /run/eve.id", false, false)
+	if err == nil {
+		fmt.Printf("  EVE-OS release %s", retStr)
+	}
+	fmt.Printf("  Edge-View Ver: %s\n", edgeViewVersion)
+	retStr, err = runCmd("uptime", false, false)
+	if err == nil {
+		ret1 := strings.Split(retStr, " load")
+		uptime := ret1[0]
+		loc, _ := time.LoadLocation("UTC")
+		fmt.Printf("  %v(%v), uptime %s", time.Now().In(loc).Format(time.RFC3339), loc, uptime)
+	}
+	fmt.Println()
+	fmt.Println()
+	closePipe(true)
+}
+
+// dynamically install package for uncommon options
+func addPackage(programName, pkgName string) error {
+	retStr, err := runCmd("which "+programName, false, false)
+	if err == nil {
+		if len(retStr) > 0 {
+			fmt.Printf("%s\n", retStr)
+		}
+	} else {
+		_, err = runCmd("apk add "+pkgName, false, false)
+		if err != nil {
+			fmt.Printf("%v\n", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func runCmd(cmd string, isEve, isPrint bool) (string, error) {
+	if isEve && !runOnServer {
+		cmd = "eve exec pillar " + cmd + " 2&>1"
+	}
+	var retStr string
+	var retBytes []byte
+	var err error
+	// The 'edge-view' can be run in 3 modes:
+	// 1) ssh-mode, from laptop using ssh into remote device for query
+	// 2) non-ssh server, the 'edge-view' is run on the device wait from client command
+	// 3) non-ssh client, the 'edge-view' is issuing command to remote 'edge-view' on device
+	//
+	// below, the case 'runOnServer' is #2 above, and else is #1, the #3 case does not come here.
+	//
+	if runOnServer {
+		// the program is running as a container on the device as a 'server',
+		// it can directly access the linux shell for requested command
+		retBytes, err = exec.Command("sh", "-c", cmd).Output()
```

Comment:
This sh -c needs to go away. Far too much security risk.
The commands should be passed as []string from the caller of runCmd and be fed into os.Exec(). That removes many of the shell tricks which can be used to launch attacks.

Reply:
For some of the 'edge-view' operations, they have grep, awk, etc things, but i'll do this:
- all the user pass in params/strings, it has to be qualified by regex allowed characters check
- we use the 'sh', '-c' cmd, only for those internal operations we know there is no problem
- for others all use []string to the os.Exec() of golang side

Reply:
@eriknordmark yes, I have limited to 'sh', '-c' to the places only necessary for those places needs them, and converted all the others to the []string into the exec(); for all the user supplied string, has to pass the regex() allowed check; I have converted all the previous coded 'cat', 'ls' to using the golang functions now.

Reply:
How about introducing specific commands for ls and cat so that we don't need the shell/ command?

Reply:
Will do. At least to start with.